flutter-engine/src/ffi.rs

```diff
@@ -175,36 +176,60 @@ impl From<FlutterPointerMouseButtons> for flutter_engine_sys::FlutterPointerMous
     }
 }
 
+pub type TextureId = i64;
+
 #[derive(Debug)]
 pub struct ExternalTexture {
-    pub(crate) engine_ptr: flutter_engine_sys::FlutterEngine,
-    pub texture_id: i64,
+    engine_ptr: flutter_engine_sys::FlutterEngine,
+    texture_id: TextureId,
 }
 
 unsafe impl Send for ExternalTexture {}
 unsafe impl Sync for ExternalTexture {}
 
-impl Drop for ExternalTexture {
-    fn drop(&mut self) {
-        trace!("dropping external texture id {}", self.texture_id);
+impl ExternalTexture {
+    pub fn new(engine_ptr: flutter_engine_sys::FlutterEngine) -> Self {
+        static TEXTURE_ID: AtomicI64 = AtomicI64::new(1);
+        let texture_id = TEXTURE_ID.fetch_add(1, Ordering::Relaxed);
+        Self {
+            engine_ptr,
+            texture_id,
+        }
+    }
+
+    pub fn id(&self) -> TextureId {
+        self.texture_id
+    }
+
+    pub fn register(&self) {
+        log::trace!("texture {}: register", self.texture_id);
         unsafe {
-            flutter_engine_sys::FlutterEngineUnregisterExternalTexture(
+            flutter_engine_sys::FlutterEngineRegisterExternalTexture(
                 self.engine_ptr,
                 self.texture_id,
             );
         }
     }
-}
 
-impl ExternalTexture {
     pub fn mark_frame_available(&self) {
+        log::trace!("texture {}: marking frame available", self.texture_id);
         unsafe {
             flutter_engine_sys::FlutterEngineMarkExternalTextureFrameAvailable(
                 self.engine_ptr,
                 self.texture_id,
             );
         }
     }
+
+    pub fn unregister(&self) {
+        log::trace!("texture {}: unregister", self.texture_id);
+        unsafe {
+            flutter_engine_sys::FlutterEngineUnregisterExternalTexture(
+                self.engine_ptr,
+                self.texture_id,
+            );
+        }
+    }
```

Comment:
Maybe this should be the `drop` handler instead?

Reply:
There is a problem that if we want to keep it `Sync` we can't have it in the drop handler, because it'll unregister when the first copy is dropped

Reply:
`Send` is probably enough, as the end user can handle the concurrency themselves, i.e. arcs?